libsolidity/analysis/ControlFlowGraph.cpp

```diff
@@ -0,0 +1,552 @@
+
+	m_currentModifierFlow.reset();
+	m_currentFunctionFlow.reset();
+
+	return false;
+}
+
+bool CFG::visit(FunctionDefinition const& _function)
+{
+	m_currentFunctionFlow = make_shared<FunctionFlow>(newNode(), newNode(), newNode());
+	m_functionControlFlow[&_function] = m_currentFunctionFlow;
+
+	ControlFlowParser::createFlowFromTo(*this, m_currentFunctionFlow->entry, m_currentFunctionFlow->exit, _function);
+
+	m_currentFunctionFlow.reset();
+
+	return false;
+}
+
+FunctionFlow const& CFG::functionFlow(FunctionDefinition const& _function) const
+{
+	solAssert(m_functionControlFlow.count(&_function), "");
+	return *m_functionControlFlow.find(&_function)->second;
+}
+
+CFGNode* CFG::newNode()
+{
+	m_nodes.emplace_back(new CFGNode());
+	return m_nodes.back().get();
+}
+
+void CFG::applyModifiers()
+{
+	for (auto const& function: m_functionControlFlow)
+	{
+		for (auto const& modifierInvocation: boost::adaptors::reverse(function.first->modifiers()))
+		{
+			if (auto modifierDefinition = dynamic_cast<ModifierDefinition const*>(
+				modifierInvocation->name()->annotation().referencedDeclaration
+			))
+			{
+				solAssert(m_modifierControlFlow.count(modifierDefinition), "");
+				auto modifierFlow = m_modifierControlFlow[modifierDefinition];
+				applyModifierFlowToFunctionFlow(*modifierFlow, function.second);
+			}
+		}
+	}
+}
+
+void CFG::applyModifierFlowToFunctionFlow(
+	ModifierFlow const& _modifierFlow,
+	std::shared_ptr<FunctionFlow> _functionFlow
+)
+{
+	map<CFGNode*, CFGNode*> oldToNew;
+
+	// inherit the revert node of the function
+	oldToNew[_modifierFlow.revert] = _functionFlow->revert;
+
+	// replace the placeholder nodes by the function entry and exit
+	oldToNew[_modifierFlow.placeholderEntry] = _functionFlow->entry;
+	oldToNew[_modifierFlow.placeholderExit] = _functionFlow->exit;
+
+	stack<CFGNode*> nodesToClone;
+	nodesToClone.push(_modifierFlow.entry);
+
+	// map the modifier entry to a new node that will become the new function entry
+	oldToNew[_modifierFlow.entry] = newNode();
+
+	while (!nodesToClone.empty())
+	{
+		auto srcNode = nodesToClone.top();
+		nodesToClone.pop();
+
+		solAssert(oldToNew.count(srcNode), "");
+
+		auto dstNode = oldToNew[srcNode];
+
+		dstNode->block = srcNode->block;
+		for (auto& entry: srcNode->entries)
+		{
+			if (!oldToNew.count(entry))
+			{
+				oldToNew[entry] = newNode();
+				nodesToClone.push(entry);
+			}
+			dstNode->entries.emplace_back(oldToNew[entry]);
+		}
+		for (auto& exit: srcNode->exits)
+		{
+			if (!oldToNew.count(exit))
+			{
+				oldToNew[exit] = newNode();
+				nodesToClone.push(exit);
+			}
+			dstNode->exits.emplace_back(oldToNew[exit]);
+		}
+	}
+
+	// if the modifier control flow never reached its exit node,
+	// we need to create a new (disconnected) exit node now
+	if (!oldToNew.count(_modifierFlow.exit))
+		oldToNew[_modifierFlow.exit] = newNode();
+
+	_functionFlow->entry = oldToNew[_modifierFlow.entry];
```

Comment:
Do the old nodes that were cloned from still exist? They are fully disconnected and could be deleted, right?

Reply:
The old nodes (i.e. the modifier nodes) aren't fully disconnected - they are unchanged.

The idea for applying a modifier to a function is to create a copy of the modifier flow and while copying to replace the placeholder entry and exit with the function's entry and exit (and also replace the revert node of the modifer with the revert node of the function).

I still need the modifier flow afterwards, since the same modifier may be applied to several functions. I'd also keep it after all modifiers have been applied, because we may want analyze modifier flow independently of function flow as well at some point.

I will try to clarify what happens by renaming the local variables here. ``oldToNew`` should probably rather be called something like ``copySrcToCopyDst`` and I'll probably not rename ``srcNode`` to ``oldNode`` and ``dstNode`` to ``newNode``, but instead ``srcNode`` to ``copySrcNode`` and ``dstNode`` to ``copyDstNode``. Do you think that would clarify what happens?

Reply:
I'll also add some documentation in the header.

Reply:
I agree about the modifier nodes, but what about the function nodes?